[clang][PowerPC] Add flag to enable compatibility with GNU for complex arguments

Long5hot opened the pull request:
Fixes : https://github.com/llvm/llvm-project/issues/56023
    
    https://godbolt.org/z/1bsW1sKMs
    
    newFlag : -fcomplex-ppc-gnu-abi
    
    GNU uses GPRs for complex parameters and return values storing for PowerPC-32bit,
    which can be enabled which above flag.
    Intent of this patch is to make clang compatible with GNU libraries of complex.

    Following up with this patch : https://reviews.llvm.org/D146942

Long5hot commented:
Following up with this patch : https://reviews.llvm.org/D146942

Long5hot commented:
ping! @chmeeedalf @nemanjai

nemanjai commented on clang/lib/CodeGen/TargetInfo.h:
```diff
@@ -486,7 +486,8 @@ std::unique_ptr<TargetCodeGenInfo>
 std::unique_ptr<TargetCodeGenInfo>
-createPPC32TargetCodeGenInfo(CodeGenModule &CGM, bool SoftFloatABI);
+createPPC32TargetCodeGenInfo(CodeGenModule &CGM, bool SoftFloatABI,
+                             unsigned RLen);
```
Please explain this change. It is not clear to me why we're adding this parameter. Are we adding support for 32-bit targets with 64-bit pointers? If so, please justify. Also, this is something that would belong in a separate patch.

Long5hot replied:
I was not sure about hardcoding Register Width(RLen) **"32"** for RLen here.. https://github.com/llvm/llvm-project/pull/77732/files#diff-05339beb4a6cf7efdfc537984255dd372dcf1edd2d00eabed25831da4d1d0a9fR276

So i used similiar approach like, how RISC-V does to calculate Register Width here.
PPC : https://github.com/llvm/llvm-project/pull/77732/files#diff-e724febedab9c1a2832bf2056d208ff02ddcb2e6f90b5a653afc9b19ac78a5d7R187
RISCV : https://github.com/llvm/llvm-project/pull/77732/files#diff-e724febedab9c1a2832bf2056d208ff02ddcb2e6f90b5a653afc9b19ac78a5d7L226

nemanjai commented:
My review is not complete, I just submitted what I have so far so at least we can get started on answering the questions I have so far.

diggerlin commented on clang/lib/CodeGen/CodeGenModule.cpp:
```diff
@@ -184,11 +184,13 @@ createTargetCodeGenInfo(CodeGenModule &CGM) {
     bool IsSoftFloat =
         CodeGenOpts.FloatABI == "soft" || Target.hasFeature("spe");
-    return createPPC32TargetCodeGenInfo(CGM, IsSoftFloat);
+    unsigned RLen = Target.getPointerWidth(LangAS::Default);
```
the variable is used only once , do not need to introduce a new variable.

diggerlin commented on clang/lib/CodeGen/CodeGenModule.cpp:
```diff
@@ -184,11 +184,13 @@ createTargetCodeGenInfo(CodeGenModule &CGM) {
     bool IsSoftFloat = CodeGenOpts.FloatABI == "soft";
-    return createPPC32TargetCodeGenInfo(CGM, IsSoftFloat);
+    unsigned RLen = Target.getPointerWidth(LangAS::Default);
+    return createPPC32TargetCodeGenInfo(CGM, IsSoftFloat, RLen);
```
the variable is used only once , do not need to introduce a new variable.

diggerlin commented on clang/lib/CodeGen/Targets/PPC.cpp:
```diff
@@ -337,12 +350,77 @@ CharUnits PPC32_SVR4_ABIInfo::getParamTypeAlignment(QualType Ty) const {
 
+ABIArgInfo PPC32_SVR4_ABIInfo::handleComplex(QualType Ty,
+                                             uint64_t &TypeSize) const {
+
```
it look the TypeSize only has 32bit and 64bit. can you change the `uint64_t &TypeSize` to `bool is64BitTypeSize`

and change following code as 

```
llvm::Type *ElemTy =is64BitTypeSize? llvm::Type::getInt64Ty(getVMContext()): llvm::Type::getInt32Ty(getVMContext());

unsigned SizeRegs = is64BitTypeSize ? 1:2;
```

diggerlin commented on clang/lib/CodeGen/Targets/PPC.cpp:
```diff
@@ -337,12 +350,77 @@ CharUnits PPC32_SVR4_ABIInfo::getParamTypeAlignment(QualType Ty) const {
+ABIArgInfo PPC32_SVR4_ABIInfo::handleComplex(QualType Ty,
+                                             uint64_t &TypeSize) const {
+
+  assert(Ty->isAnyComplexType());
```
it looks the Ty pass in the function is only used for assert ?

diggerlin commented on clang/lib/CodeGen/Targets/PPC.cpp:
```diff
@@ -337,12 +350,77 @@ CharUnits PPC32_SVR4_ABIInfo::getParamTypeAlignment(QualType Ty) const {
+    ElemTy = llvm::Type::getInt64Ty(getVMContext());
+    SizeRegs = 1;
+  } else {
+    ElemTy = llvm::Type::getInt32Ty(getVMContext());
```
I am curiously , why not get ElemTy = llvm::Type::getInt128Ty(getVMContext()); instead of ElemTy = llvm::Type::getInt32Ty(getVMContext());

Long5hot replied:
> llvm::Type *ElemTy = llvm::Type::getInt64Ty(getVMContext());
> unsigned SizeRegs = TypeSize/64;


Since Typesize will be 64, in if is it necessary to replace 1 to TypeSize/64?

diggerlin replied:
got it, thanks for explain. can you add your explain as comment ?

diggerlin commented on clang/lib/CodeGen/Targets/PPC.cpp:
```diff
@@ -337,12 +350,77 @@ CharUnits PPC32_SVR4_ABIInfo::getParamTypeAlignment(QualType Ty) const {
+    if (TypeSize > RLen && TypeSize <= 2 * RLen)
+      ArgGPRsLeft -= 2;
+    else
+      ArgGPRsLeft--;
```
There is a bug here, you decrease the ArgGPRsLeft no mater whether ArgGPRsLeft >0;

I think you can put after `Ty = useFirstFieldIfTransparentUnion(Ty);`
```
Ty = useFirstFieldIfTransparentUnion(Ty);
if(Ty->isFloatingType()) return;
```
and change the code here as 
```
if(TypeSize <= RLen * ArgGPRsLeft) 
    ArgGPRsLeft -= TypeSize / RLen;
```

amy-kwan commented on clang/lib/CodeGen/Targets/PPC.cpp:
```diff
@@ -337,12 +350,77 @@ CharUnits PPC32_SVR4_ABIInfo::getParamTypeAlignment(QualType Ty) const {
+ABIArgInfo PPC32_SVR4_ABIInfo::handleComplex(QualType Ty,
+                                             uint64_t &TypeSize) const {
+
+  assert(Ty->isAnyComplexType());
```
It would also be good if the assert has a message.
Additionally, for people who do not do assert builds, since `Ty` is only used for the assert, they may get an unused variable warning.

diggerlin replied:
if use only for the assert, suggest putting assert outside the function,

amy-kwan commented on clang/lib/CodeGen/Targets/PPC.cpp:
```diff
@@ -361,6 +439,9 @@ ABIArgInfo PPC32_SVR4_ABIInfo::classifyReturnType(QualType RetTy) const {
       return ABIArgInfo::getDirect(CoerceTy);
     }
   }
+  if (isComplexInRegABI && RetTy->isAnyComplexType()) {
```
Braces not needed here.

amy-kwan reviewed (commented):
I think the braces can also be omitted on the conditions within the `clang/*` files.

diggerlin commented on clang/lib/CodeGen/Targets/PPC.cpp:
```diff
@@ -271,22 +271,33 @@ namespace {
       FI.getReturnInfo() = classifyReturnType(FI.getReturnType());
     for (auto &I : FI.arguments())
-      I.info = classifyArgumentType(I.type);
+      I.info = classifyArgumentType(I.type, ArgGPRsLeft);
```
why do you need  introduce a new variable `ArgGPRsLeft` since you do not use the new variable after the call .

diggerlin commented on clang/lib/CodeGen/Targets/PPC.cpp:
```diff
@@ -372,11 +453,12 @@ Address PPC32_SVR4_ABIInfo::EmitVAArg(CodeGenFunction &CGF, Address VAList,
     return emitVoidPtrVAArg(CGF, VAList, Ty,
-                            classifyArgumentType(Ty).isIndirect(), TI, SlotSize,
-                            /*AllowHigherAlign=*/true);
+                            classifyArgumentType(Ty, ArgGPRs).isIndirect(), TI,
```
I am sure purpose the calculation the variable `ArgGPRs`, it be calculated in the function  `classifyArgumentType` , but never use it.

diggerlin commented on clang/include/clang/Basic/CodeGenOptions.h:
```diff
@@ -78,6 +78,13 @@ class CodeGenOptions : public CodeGenOptionsBase {
 
+  enum ComplexArgumentConventionKind {
+    CMPLX_Default,
+    CMPLX_OnStack,
```
it looks you define the `CMPLX_Default` and   ` CMPLX_OnStack` , but never use them.

Long5hot replied:
This was for future reference, if any arch wants to use this.

efriedma-quic replied:
I don't think this needs to be a target-independent thing; this doesn't really generalize to other targets.  I mean, maybe different targets have sort of similar schemes, but the same ABI compatibility issue shouldn't exist anywhere else.  Maybe pass this as a target feature (ref getPPCTargetFeatures()).

diggerlin commented on clang/lib/CodeGen/Targets/PPC.cpp:
```diff
@@ -271,22 +271,33 @@ namespace {
 
   void computeInfo(CGFunctionInfo &FI) const override {
+
+    int ArgGPRsLeft = NumArgGPRs;
```
nit: move the definition to the closest first be used.

Long5hot replied:
Couldn't understand what you mean here..

diggerlin replied:
put `int ArgGPRsLeft = NumArgGPRs;` in front of 

 `   for (auto &I : FI.arguments())
      I.info = classifyArgumentType(I.type, ArgGPRsLeft);`

diggerlin commented on clang/lib/CodeGen/Targets/PPC.cpp:
```diff
@@ -337,12 +350,77 @@ CharUnits PPC32_SVR4_ABIInfo::getParamTypeAlignment(QualType Ty) const {
+    return handleComplex(Ty, TypeSize);
+  }
+
+  if (isAggregateTypeForABI(Ty)) {
```
it look, you only has Complex as parameter in your test case .for example `_Complex float foo1(_Complex float x)`

I think you need more test case with other parameters to test the functionality of function `classifyArgumentType()`

Long5hot commented on clang/lib/CodeGen/Targets/PPC.cpp:
```diff
@@ -271,22 +271,32 @@ namespace {
+  bool IsComplexInRegABI;
+  // Size of GPR in bits.
+  static const unsigned RegLen = 32;
+  static const int NumArgGPRs = 8;
```
The purpose of NumArgGPRs or ArgGPRsLeft is limited to complex types.

diggerlin commented on clang/lib/CodeGen/Targets/PPC.cpp:
```diff
@@ -337,12 +347,58 @@ CharUnits PPC32_SVR4_ABIInfo::getParamTypeAlignment(QualType Ty) const {
+    else if (!Ty->isFloatingType()) {
+      // For other premitive types.
+      if (TypeSize > RegLen && TypeSize <= 2 * RegLen)
+        ArgGPRsLeft -= TypeSize / RegLen;
```
can we change 
```
if (TypeSize > RegLen && TypeSize <= 2 * RegLen)
        ArgGPRsLeft -= TypeSize / RegLen;
```

and what if ArgGPRsLeft =1 and  TypeSize == 2 * RegLen. after above code , the ArgGPRsLeft  will be -1. is it correct ? it hit assert `assert(ArgGPRsLeft <= NumArgGPRs && "Arg GPR tracking underflow");` when the `classifyArgumentType` is called next time.

suggest change to 
 ```
 if (TypeSize > RegLen &&  && TypeSize <= RLen * ArgGPRsLeft)
        ArgGPRsLeft -= TypeSize / RegLen;
```

and add a test case which can make ArgGPRsLeft  to zero;

diggerlin replied:
and If I understand correct, the single precision floating point complex . and 64 bit integer (long long unsigned)  also need a  even alignment .

`If gr > 9, go to OTHER. If gr is even, set gr = gr + 1. Load the lower-addressed word of the
argument into gr and the higher-addressed word into gr + 1, set gr = gr + 2,` 

it looks  the logic not express in your code when you calculate the `ArgGPRsLeft` ?

Long5hot replied:
llvm powerpc backend takes care of that. 
refer CC_PPC32_SVR4_Custom_AlignArgRegs method from PPCCallingconvention.cpp https://github.com/llvm/llvm-project/blob/main/llvm/lib/Target/PowerPC/PPCCallingConv.cpp#L80

diggerlin replied:
for a test case `foo6(_Complex long double foo4(int x1, long  long y1, int x2, _Complex double z,)`
 
 `int x1` will have r3, 
`long long y1` will have r5,r6 (skip r4 since it is even, for 64bit integer need to start start odd register.)
`int x2` will have r7, 
` _Complex double z`  need to occupy 4 registers, there is no enough register , it can not pass `_Complex double z` in r8,r9,r10 , you have to pass ` _Complex double z` in by pointer (you can not call `handleComplex()`). 

but in your code, 
ArgGPRsLeft is 7 after int x1,
ArgGPRsLeft is 5 after long long y1
ArgGPRsLeft is 4 after int x2

there is still have 4 registers left for ` _Complex double z`, you can call `handleComplex()`,  you will pass ` _Complex double z` by value 

can you check how gcc deal with this case   `foo6(_Complex long double foo4(int x1, long  long y1, int x2, _Complex double z,)`  ?

diggerlin replied:
for a test case `foo6(_Complex long double foo4(int x1, long  long y1, int x2, _Complex double z,)`
 
 `int x1` will have r3, 
`long long y1` will have r5,r6 (skip r4 since it is even, for 64bit integer need to start start odd register.)
`int x2` will have r7, 
` _Complex double z`  need to occupy 4 registers, there is no enough register , it can not pass `_Complex double z` in r8,r9,r10 , you have to pass ` _Complex double z` in by pointer (you can not call `handleComplex()`). 

but in your code, 
ArgGPRsLeft is 7 after int x1,
ArgGPRsLeft is 5 after long long y1
ArgGPRsLeft is 4 after int x2

there is still have 4 registers left for ` _Complex double z`, you can call `handleComplex()`,  you will pass ` _Complex double z` by value 

can you check how gcc deal with this case   `foo6(_Complex long double foo4(int x1, long  long y1, int x2, _Complex double z,)`  ?

Long5hot replied:
Yes, gcc  passes the last arg in stack
https://godbolt.org/z/EbPMK4jad

diggerlin commented on clang/lib/CodeGen/Targets/PPC.cpp:
```diff
@@ -337,12 +347,58 @@ CharUnits PPC32_SVR4_ABIInfo::getParamTypeAlignment(QualType Ty) const {
+    RegsNeeded = 1;
+  } else {
+    ElemTy = llvm::Type::getInt32Ty(getVMContext());
+    RegsNeeded = (TypeSize + 31) / 32;
```
do we need to add `+31` here , can we change to 
`RegsNeeded = TypeSize >> 5;  ` ? and

diggerlin commented on clang/lib/CodeGen/Targets/PPC.cpp:
```diff
@@ -337,12 +347,58 @@ CharUnits PPC32_SVR4_ABIInfo::getParamTypeAlignment(QualType Ty) const {
+
+ABIArgInfo PPC32_SVR4_ABIInfo::classifyArgumentType(QualType Ty,
+                                                    int &ArgGPRsLeft) const {
+  assert(ArgGPRsLeft <= NumArgGPRs && "Arg GPR tracking underflow");
```
and add assert(ArgGPRsLeft >= 0 && "Arg GPR must be large or equal than zero"); ?

diggerlin commented on clang/test/CodeGen/PowerPC/ppc32-complex-gnu-abi.c:
```diff
@@ -0,0 +1,177 @@
+// RUN:   -emit-llvm %s -o - | FileCheck %s --check-prefix=CHECK-GNU
+
+// CHECK-DEF-LABEL: define dso_local void @foo1
+// CHECK-DEF-SAME: (ptr dead_on_unwind noalias writable sret({ float, float }) align 4 [[AGG_RESULT:%.*]], ptr noundef byval({ float, float }) align 4 [[X:%.*]]) #[[ATTR0:[0-9]+]] {
```
nit, we can delete  `#[[ATTR0:[0-9]+]]1` , and please delete other place as well.

Long5hot replied:
I couldn't remove #[[ATTR0:[0-9]+]]1. Because testcase fails saying no missing attributes.
llvm-lit(clang) generates IR with # attributes, so looks like Filecheck expects #attribute fields to be present.

diggerlin commented on clang/lib/CodeGen/Targets/PPC.cpp:
```diff
@@ -396,12 +405,85 @@ CharUnits PPC32_SVR4_ABIInfo::getParamTypeAlignment(QualType Ty) const {
+        ArgGPRsLeft -= TypeSize / RegLen;
+      }
+    } else
+      --ArgGPRsLeft;
```
it look that if TypeSize=128, you only do `--ArgGPRsLeft`,  it pass by pointer ?

Long5hot replied:
I missed the part for soft-float, for soft-float we pass long-double(128) in GPRs, Below change works.

```
else if (!Ty->isFloatingType() || (Ty->isFloatingType() && IsSoftFloatABI)) { 
if (TypeSize == 64 && ArgGPRsLeft % 2 == 1)
     --ArgGPRsLeft; // If gr is even set gr = gr + 1 for TypeSize=64.
if (TypeSize <= ArgGPRsLeft * RegLen)
     ArgGPRsLeft -= TypeSize / RegLen;
   }
```

I will update the review as your comments. Do i need to add the tests for soft-floats as well??

diggerlin replied:
it is nice to have.

diggerlin commented on clang/lib/CodeGen/Targets/PPC.cpp:
```diff
@@ -421,6 +503,11 @@ ABIArgInfo PPC32_SVR4_ABIInfo::classifyReturnType(QualType RetTy) const {
 
+  bool IsComplexInRegABI =
+      getCodeGenOpts().getComplexInRegABI() == CodeGenOptions::CMPLX_InGPR;
+  if (IsComplexInRegABI && RetTy->isAnyComplexType())
```
you do not need to introduce a new variable `IsComplexInRegABI` here since it only be used once.

change to 
`if ((getCodeGenOpts().getComplexInRegABI() == CodeGenOptions::CMPLX_InGPR) && RetTy->isAnyComplexType())`

diggerlin commented on clang/test/CodeGen/PowerPC/ppc32-complex-gnu-abi.c:
```diff
@@ -0,0 +1,306 @@
+// CHECK-GNU-NEXT:    [[TMP0:%.*]] = load [1 x i64], ptr [[RETVAL]], align 4
+// CHECK-GNU-NEXT:    ret [1 x i64] [[TMP0]]
+//
+_Complex float foo1(_Complex float x) {
```
general question for all test scenario: 

can you add a caller function to call the function to check whether organize the call parameter and pass the parameter as called function need ?

diggerlin commented on clang/test/CodeGen/PowerPC/ppc32-complex-soft-float-gnu-abi.c:
```diff
@@ -0,0 +1,350 @@
+// CHECK-GNU-SOFT-FLOAT-NEXT:    [[TMP0:%.*]] = load [4 x i32], ptr [[RETVAL]], align 8
+// CHECK-GNU-SOFT-FLOAT-NEXT:    ret [4 x i32] [[TMP0]]
+//
+_Complex double _cdouble(_Complex double x) {
```
there is not float or double as parameter in the test case , I think we have tested the  scenario in the ppc32-complex-gnu-abi.c 
, do we still need it ?

diggerlin commented on clang/test/CodeGen/PowerPC/ppc32-complex-soft-float-gnu-abi.c:
```diff
@@ -0,0 +1,350 @@
+// CHECK-GNU-SOFT-FLOAT-NEXT:    [[TMP0:%.*]] = load [8 x i32], ptr [[RETVAL]], align 16
+// CHECK-GNU-SOFT-FLOAT-NEXT:    ret [8 x i32] [[TMP0]]
+//
+_Complex long double _cldouble(float f, _Complex long double x) {
```
do you need the test scenario ? 

I think we only test case which have float or double as parameter and make a _Complex parameter is passed as pointer because option  '-msoft-float` is enabled

Long5hot replied:
TestCase is same as above, just replacing integer premitives with floating point, Considering this because for soft float also get's passed in gprs..

diggerlin commented on clang/test/CodeGen/PowerPC/ppc32-complex-soft-float-gnu-abi.c:
```diff
@@ -0,0 +1,286 @@
+// CHECK-GNU-DEF-NEXT:    ret [8 x i32] [[TMP0]]
+//
+// CHECK-GNU-SOFT-FLOAT-LABEL: define dso_local [8 x i32] @_cldouble
+// CHECK-GNU-SOFT-FLOAT-SAME: (float noundef [[F:%.*]], ptr noundef byval({ ppc_fp128, ppc_fp128 }) align 16 [[X:%.*]]) #[[ATTR0:[0-9]+]] {
```
I am curiously that option `-msoft-float` do not change the `float noundef [[F:%.*]]` to `i32 noundef [[F:%.*]]` , I do not read the code implement of the option `-msoft-float`

Long5hot replied:
Below assembly's generated with -fcomplex-ppc-gnu-abi -O1 --target=ppc32 -S  -msoft-float

"use-soft-float"="true" get's passed in IR in attributes

```
_Complex double testComplexDouble(float w, _Complex float x, _Complex double z)
{
  return z;
}

testComplexDouble:                      # @testComplexDouble
.Lfunc_begin2:
        .cfi_startproc
# %bb.0:
        mr      6, 10
        mr      5, 9
        mr      4, 8
        mr      3, 7
        blr
.Lfunc_end2:
        .size   testComplexDouble, .Lfunc_end2-.Lfunc_begin2
        .cfi_endproc
                                        # -- End function
```

```
_Complex double checkComplexDoubleOnStack(float x1, _Complex float cf, float x2, _Complex double cd)
{
  return testComplexDouble(x2, cf, cd);
}

        .globl  checkComplexDoubleOnStack       # -- Begin function checkComplexDoubleOnStack
        .p2align        2
        .type   checkComplexDoubleOnStack,@function
checkComplexDoubleOnStack:              # @checkComplexDoubleOnStack
.Lfunc_begin3:
        .cfi_startproc
# %bb.0:
        lwz 3, 0(8)
        lwz 4, 4(8)
        lwz 5, 8(8)
        lwz 6, 12(8)
        blr
```

diggerlin commented on clang/test/CodeGen/PowerPC/ppc32-complex-soft-float-gnu-abi.c:
```diff
@@ -0,0 +1,286 @@
+// CHECK-GNU-DEF-NEXT:    store double [[Z_IMAG]], ptr [[RETVAL_IMAGP]], align 8
+// CHECK-GNU-DEF-NEXT:    [[TMP0:%.*]] = load [4 x i32], ptr [[RETVAL]], align 8
+// CHECK-GNU-DEF-NEXT:    ret [4 x i32] [[TMP0]]
+//
```
nit : using empty line instead using this `//`  line,  please do for all of them.

diggerlin commented on clang/test/CodeGen/PowerPC/ppc32-complex-soft-float-gnu-abi.c:
```diff
@@ -0,0 +1,286 @@
+// CHECK-GNU-SOFT-FLOAT-NEXT:    [[TMP3:%.*]] = load [4 x i32], ptr [[RETVAL]], align 8
+// CHECK-GNU-SOFT-FLOAT-NEXT:    ret [4 x i32] [[TMP3]]
+//
+_Complex double checkComplexDoubleOnStack(float x1, _Complex float cf, float x2, _Complex double cd)
```
I think we only the test senario is enough to test the `float` parameter occups one gpr in the `PPC32_SVR4_ABIInfo::classifyArgumentType`. We can delete all other scenarios.

diggerlin commented on clang/test/Driver/ppc32-fcomplex-ppc-gnu-abi.c:
```diff
@@ -0,0 +1,15 @@
+// RUN:     | FileCheck %s -check-prefix=X86_64
+// X86_64: error: unsupported option '-fcomplex-ppc-gnu-abi' for target 'x86_64'
+
+// RUN: not %clang %s --target=ppc64 -fcomplex-ppc-gnu-abi 2>&1 \
```
I think two test scenarios is enough , 

`if (!TC.getTriple().isPPC32() || !TC.getTriple().isOSBinFormatELF()) {`

one for `!TC.getTriple().isPPC32()` , other for `!TC.getTriple().isOSBinFormatELF()`

diggerlin commented:
please do not merge the commit if possible, merging the commit make the review difficult,  We want to compare what change between the two commit when we review. otherwise we have review the patch from scratch.

diggerlin commented:
I do not further comment on the PR, 

but I have worry about that , the PR have different mechanism with "-msoft-float" , the -msoft-float pass "use-soft-float"="true" in the IR, the llc will l put float parameter in the GPR, but the PR implement `-fcomplex-ppc-gnu-abi` in the frontend  , can we implement  `-fcomplex-ppc-gnu-abi`  as  "-msoft-float"   ?

UmeshKalappa0 commented:
@diggerlin  ,
>>can we implement -fcomplex-ppc-gnu-abi as "-msoft-float" ? what is your opinion ?
that was  not the intent here and we need to consultant ABI reference regrading the same ,before we decide on implementation .

Long5hot commented:
ping!! @nemanjai

Long5hot commented:
Ping!!

Long5hot commented:
ping @chmeeedalf @daltenty @diggerlin @amy-kwan !!

Long5hot commented:
ping!

AaronBallman commented on clang/lib/Frontend/CompilerInvocation.cpp:
```diff
@@ -1654,6 +1654,10 @@ void CompilerInvocationBase::GenerateCodeGenArgs(const CodeGenOptions &Opts,
     GenerateArg(Consumer, Opt);
   }
 
+  if (T.isPPC32() && Opts.ComplexInRegABI == CodeGenOptions::CMPLX_InGPR) {
+    GenerateArg(Consumer, OPT_fcomplex_ppc_gnu_abi);
+  }
+
```
```suggestion
  if (T.isPPC32() && Opts.ComplexInRegABI == CodeGenOptions::CMPLX_InGPR)
    GenerateArg(Consumer, OPT_fcomplex_ppc_gnu_abi);

```

AaronBallman commented on clang/lib/Frontend/CompilerInvocation.cpp:
```diff
@@ -2028,6 +2032,10 @@ bool CompilerInvocation::ParseCodeGenArgs(CodeGenOptions &Opts, ArgList &Args,
     }
   }
 
+  if (Args.getLastArg(OPT_fcomplex_ppc_gnu_abi)) {
+    Opts.setComplexInRegABI(CodeGenOptions::CMPLX_InGPR);
+  }
```
```suggestion
  if (Args.getLastArg(OPT_fcomplex_ppc_gnu_abi))
    Opts.setComplexInRegABI(CodeGenOptions::CMPLX_InGPR);
```

AaronBallman commented on clang/test/Driver/ppc32-fcomplex-ppc-gnu-abi.c:
```diff
@@ -0,0 +1,9 @@
+// RUN: %clang -c --target=ppc32 -fcomplex-ppc-gnu-abi %s 2>&1
```
It'd be nice for this to check `-###` output to ensure that it's passing along the correct `-cc1` option.

AaronBallman reviewed (commented):
I'm wondering why the GNU-compatible behavior is not on by default, especially when in GNU mode (e.g., `-std=gnu17`)?

Also, the changes should come with a release note in `clang/docs/ReleaseNotes.rst` so users know about the new functionality.

MaskRay commented on clang/test/CodeGen/PowerPC/ppc32-complex-gnu-abi.c:
```diff
@@ -0,0 +1,396 @@
+// RUN:   -emit-llvm %s -o - | FileCheck %s --check-prefix=CHECK-DEF
+// RUN: %clang_cc1 -triple powerpc-unknown-linux-gnu -fcomplex-ppc-gnu-abi \
+// RUN:   -emit-llvm %s -o - | FileCheck %s --check-prefix=CHECK-GNU
+
```
We just use one single blank line as a separator.

AaronBallman commented:
> > I'm wondering why the GNU-compatible behavior is not on by default, especially when in GNU mode (e.g., `-std=gnu17`)?
> 
> ATR-PASS-COMPLEX-IN-GPRS was missing in clang as shown in ABI doc : https://example61560.wordpress.com/wp-content/uploads/2016/11/powerpc_abi.pdf

I'm sorry but I still don't understand why it's not on by default in GNU mode. If the user is in GNU mode, they presumably are doing so because they want to be GNU-compatible, yes?

> > Also, the changes should come with a release note in `clang/docs/ReleaseNotes.rst` so users know about the new functionality.
> 
> @AaronBallman, can you guide me, in which section of ReleaseNotes.rst i should add info about new flag?

I'd add it under `New Compiler Flags`

Long5hot commented:
We were compiling simple testcase with complex parameters with clang and it was crashing because libraries was built using gcc.

@AaronBallman, Reason for new flag was to enable this for other C standards as well. Currently we use c11 as standard in VxWorks toolchain, which has to work with gcc compiled libraries.
gcc with c11 as well passes complex parameters in GPRs.

AaronBallman commented:
> We were compiling simple testcase with complex parameters with clang and it was crashing because libraries was built using gcc.
> 
> @AaronBallman, Reason for new flag was to enable this for other C standards as well. Currently we use c11 as standard in VxWorks toolchain, which has to work with gcc compiled libraries. gcc with c11 as well passes complex parameters in GPRs.

Ah, perhaps I was unclear -- I'm not opposed to the flag existing so users can opt into/out of it as they want. I'm wondering why the flag needs to be specified when compiling with `-triple powerpc-unknown-linux-gnu` (because that specifies a GNU environment with a powerpc target) or with `-triple powerpc -std=gnu<whatever>` (because that specifies GNU extensions with a powerpc target). It seems to me that the flag should be implied in those cases, so users shouldn't have to specify it manually, right?

Long5hot commented:
@AaronBallman, Yes you are right. I will work on it!

Long5hot committed:
[clang][PowerPC] Add flag to enable compatibility with GNU for complex arguments (#77732)

Fixes : https://github.com/llvm/llvm-project/issues/56023

https://godbolt.org/z/1bsW1sKMs

newFlag : -fcomplex-ppc-gnu-abi

GNU uses GPRs for complex parameters and return values storing for PowerPC-32bit,
which can be enabled which above flag.
Intent of this patch is to make clang compatible with GNU libraries of complex.

Following up with this patch : https://reviews.llvm.org/D146942

Long5hot commented:
I'm able to query if triple has gnu but can't query for -std=gnu..!!

efriedma-quic commented:
`-std=gnu99` etc. sets LangOpts.GNUMode... but we don't want to base the ABI on that; it's not an ABI flag.  The ABI should only be based on the triple, and flags specifically intended to modify the ABI.

I think we should just use the GNU-compatible ABI by default.  Unless there's some existing OS where clang is the platform compiler that wants to preserve the old ABI, but even if such a platform exists, I doubt there's anyone we can ask.

brad0 commented:
cc @kernigh

Long5hot commented:
@efriedma-quic for the same reason, we created this switch. There was a debate regarding, reason being we concluded to creating the flag and not choosing GNU by default.
So now should we go ahead and push this patch?

Long5hot commented:
ping!

kernigh commented:
I am distracted by issue #133507 (clang-19 miscompiles itself for PPC32), so I can't check this pull request right now. [My comment of August 29, 2023](https://reviews.llvm.org/D146942#4626773) had this example:

```
#include "complex.h"

double complex
third(int no, double complex no_again, double complex yes)
{
    return yes;
}
```

gcc passed _yes_ on the stack, but clang with the 2023 patch split _yes_ between registers and the stack, so it was not yet compatible. Has the patch been fixed? I don't want clang to invent a 3rd complex ABI that is incompatible with both gcc and old clang.

> Unless there's some existing OS where clang is the platform compiler that wants to preserve the old ABI, but even if such a platform exists, I doubt there's anyone we can ask.

For 32-bit PowerPC, I believe that clang can target Linux, NetBSD, FreeBSD, and OpenBSD. There are complex math functions like [csqrt(3)](https://man.openbsd.org/csqrt.3). I believe that Linux and NetBSD compile these functions with gcc, while FreeBSD and OpenBSD compile them with clang. For OpenBSD, talk to me. For FreeBSD, I don't know.

I might want OpenBSD to switch back to gcc's ABI. OpenBSD 6.7 (in 2020) had changed its powerpc compiler from gcc to clang. This was before issue #56023 (in 2022) reported the ABI change. Back in 2020, nobody noticed OpenBSD's ABI change, but now, if I use gcc, I can't call csqrt(3). If clang switched to gcc's ABI, then gcc would work again.

Long5hot commented:
> Has the patch been fixed

Yess, Works expected as gnu!

efriedma-quic reviewed (commented):
Being incompatible with the platform compiler is the wrong default; we should be compatible by default.  And the platform compiler is gcc for almost actively developed ppc32 targets, so we should just switch the default.  Having a flag in addition to that seems fine if you think there's a chance people will need to control it.
